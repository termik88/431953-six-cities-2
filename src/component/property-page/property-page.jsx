import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {transformRatingToStarsNumber, getPlacesNearest} from "../../until.js";

import {getAuthorizationStatus} from "../../reducer/user/selector.js";
import {getPlacesAll, getComments, getErrorInfo, getIsLoading} from "../../reducer/data/selectors.js";
import {Operations} from "../../reducer/data/data.js";

import Review from "../review/review.jsx";
import Card from "../card/card.jsx";
import Map from "../map/map.jsx";
import {FavoriteButtonContainer} from "../favorite-button/favorite-button.jsx";

const IMAGES_MAX = 6;

class PropertyPage extends PureComponent {
  componentDidMount() {
    this.props.loadDataComments(this.props.match.params.id);
  }

  render() {
    const placeCurrent = this.props.places.find((item) => item.id === +this.props.match.params.id);
    const placesNearest = getPlacesNearest(placeCurrent, this.props.places);

    const {
      id,
      images,
      title,
      isFavorite,
      isPremium,
      bedrooms,
      maxAdults,
      rating,
      price,
      goods,
      type,
      host,
      description
    } = placeCurrent;
    return (
      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">

              {images.slice(0, IMAGES_MAX).map((it, i) =>
                <div key={`image-${i}`} className="property__image-wrapper">
                  <img className="property__image" src={it} alt="Photo studio"/>
                </div>)}

            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {isPremium && (
                <div className="property__mark">
                  <span>Premium</span>
                </div>
              )}
              <div className="property__name-wrapper">
                <h1 className="property__name">{title}</h1>

                <FavoriteButtonContainer
                  buttonName = {`property`}
                  id = {id}
                  isFavorite = {isFavorite}
                />

              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `${rating}%`}}/>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{transformRatingToStarsNumber(rating)}</span>
              </div>

              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {bedrooms} Bedrooms
                </li>
                <li className="property__feature property__feature--adults">
                  Max {maxAdults} adults
                </li>
              </ul>

              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((it, i) => <li key={i} className="property__inside-item"> {it} </li>)}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div key={host.id} className="property__host-user user">
                  <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                    <img
                      className="property__avatar user__avatar"
                      src={`../${host.avatarUrl}`}
                      width="74"
                      height="74"
                      alt="Host avatar"/>
                  </div>
                  <span className="property__user-name">{host.name}</span>
                  {host.isPro && (<span className="property__user-status">
                  Pro
                  </span>)}
                </div>
                <div className="property__description">
                  <p className="property__text">
                    {description}
                  </p>
                </div>
              </div>

              <Review
                placeId = {id}
                comments = {this.props.comments}
                isAuthorizationRequired = {this.props.isAuthorizationRequired}
                onSendComment = {this.props.onSendComment}
                isLoading = {this.props.isLoading}
                errorInfo = {this.props.errorInfo}
              />

            </div>
          </div>

          <section className="property__map map">
            <Map
              places = {placesNearest}
              active = {placeCurrent}
            />
          </section>

        </section>

        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <div className="near-places__list places__list">

              {placesNearest.map((item) => (
                <Card
                  key = {`card-${item.id}`}
                  cardName = {`near`}
                  place = {item}
                />
              ))}

            </div>
          </section>
        </div>

      </main>
    );
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.id !== prevProps.match.params.id) {
      this.props.loadDataComments(this.props.match.params.id);
    }
  }
}

PropertyPage.propTypes = {
  places: PropTypes.arrayOf(PropTypes.exact({
    id: PropTypes.number,
    city: PropTypes.exact({
      name: PropTypes.string,
      location: PropTypes.exact({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number
      })
    }),
    previewImage: PropTypes.string,
    images: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string,
    isFavorite: PropTypes.bool,
    isPremium: PropTypes.bool,
    rating: PropTypes.number,
    type: PropTypes.string,
    bedrooms: PropTypes.number,
    maxAdults: PropTypes.number,
    price: PropTypes.number,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.exact({
      id: PropTypes.number,
      isPro: PropTypes.bool,
      name: PropTypes.string,
      avatarUrl: PropTypes.string
    }),
    description: PropTypes.string,
    location: PropTypes.exact({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number
    })
  })).isRequired,
  comments: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      avatarUrl: PropTypes.string.isRequired,
    }).isRequired,
    rating: PropTypes.number.isRequired,
    comment: PropTypes.string,
    date: PropTypes.string.isRequired,
  })),
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    })
  }),
  isAuthorizationRequired: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  errorInfo: PropTypes.string.isRequired,
  onSendComment: PropTypes.func.isRequired,
  loadDataComments: PropTypes.func.isRequired,

};

const mapStateToProps = (state) => ({
  places: getPlacesAll(state),
  comments: getComments(state),
  isAuthorizationRequired: getAuthorizationStatus(state),
  isLoading: getIsLoading(state),
  errorInfo: getErrorInfo(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadDataComments: (id) => dispatch(Operations.loadDataComments(id)),
  onSendComment: (id, comment, callBack) => dispatch(Operations.sendComment(id, comment, callBack))
});

const PropertyPageContainer = connect(mapStateToProps, mapDispatchToProps)(PropertyPage);

export {PropertyPage, PropertyPageContainer};
