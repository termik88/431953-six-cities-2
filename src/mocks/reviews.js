const reviews = [
  {
    id: 1,
    user: {
      id: 15,
      name: `Kendall`,
      avatarUrl: `https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/6.jpg`,
      isPro: false
    },
    rating: 80,
    date: `2019-12-04T09:09:49.938Z`,
    commentText: `We loved it so much, the house, the veiw, the location just great.. Highly reccomend :)`
  },
  {
    id: 2,
    user: {
      id: 14,
      name: `Corey`,
      avatarUrl: `https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/5.jpg`,
      isPro: true
    },
    rating: 60,
    date: `2019-12-04T09:09:49.938Z`,
    commentText: `What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!`
  },
  {
    id: 3,
    user: {
      id: 19,
      name: `Christina`,
      avatarUrl: `https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/10.jpg`,
      isPro: false
    },
    rating: 60,
    date: `2019-12-04T09:09:49.938Z`,
    commentText: `What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!`
  }
];

export const userData = {
  id: 1,
  email: `Oliver.conner@gmail.com`,
  name: `Oliver.conner`,
  avatarUrl: `img/1.png`,
  isPro: false
};

export default reviews;
