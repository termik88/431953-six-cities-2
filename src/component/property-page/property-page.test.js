import React from 'react';
import renderer from 'react-test-renderer';

import places from "../../mocks/places.js";
import reviews from "../../mocks/reviews.js";

import {PropertyPage} from "./property-page.jsx";

jest.mock(`../favorite-button/favorite-button.jsx`);
jest.mock(`../map/map.jsx`);

describe(`PropertyPage component test.`, () => {
  it(`Correctly renders after relaunch`, () => {
    const mockDate = {
      places,
      comments: reviews,
      match: {
        params: {
          id: `1`
        }
      },
      isAuthorizationRequired: false,
      isLoading: false,
      errorInfo: ``,
      onSendComment: jest.fn(),
      loadDataComments: jest.fn(),
    };

    const component = renderer
      .create(
          <PropertyPage
            {...mockDate}
          />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
