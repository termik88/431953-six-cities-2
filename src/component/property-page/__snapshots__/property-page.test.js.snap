// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyPage component test. Correctly renders after relaunch 1`] = `
<main
  className="page__main page__main--property"
>
  <section
    className="property"
  >
    <div
      className="property__gallery-container container"
    >
      <div
        className="property__gallery"
      >
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg"
          />
        </div>
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg"
          />
        </div>
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg"
          />
        </div>
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg"
          />
        </div>
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg"
          />
        </div>
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg"
          />
        </div>
      </div>
    </div>
    <div
      className="property__container container"
    >
      <div
        className="property__wrapper"
      >
        <div
          className="property__name-wrapper"
        >
          <h1
            className="property__name"
          >
            Canal View Prinsengracht
          </h1>
        </div>
        <div
          className="property__rating rating"
        >
          <div
            className="property__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "3.1%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
          <span
            className="property__rating-value rating__value"
          >
            0.155
          </span>
        </div>
        <ul
          className="property__features"
        >
          <li
            className="property__feature property__feature--entire"
          >
            room
          </li>
          <li
            className="property__feature property__feature--bedrooms"
          >
            1
             Bedrooms
          </li>
          <li
            className="property__feature property__feature--adults"
          >
            Max 
            2
             adults
          </li>
        </ul>
        <div
          className="property__price"
        >
          <b
            className="property__price-value"
          >
            €
            101
          </b>
          <span
            className="property__price-text"
          >
             night
          </span>
        </div>
        <div
          className="property__inside"
        >
          <h2
            className="property__inside-title"
          >
            What's inside
          </h2>
          <ul
            className="property__inside-list"
          >
            <li
              className="property__inside-item"
            >
               
              Breakfast
               
            </li>
            <li
              className="property__inside-item"
            >
               
              Laptop friendly workspace
               
            </li>
          </ul>
        </div>
        <div
          className="property__host"
        >
          <h2
            className="property__host-title"
          >
            Meet the host
          </h2>
          <div
            className="property__host-user user"
          >
            <div
              className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
            >
              <img
                alt="Host avatar"
                className="property__avatar user__avatar"
                height="74"
                src="../img/avatar-angelina.jpg"
                width="74"
              />
            </div>
            <span
              className="property__user-name"
            >
              Angelina
            </span>
            <span
              className="property__user-status"
            >
              Pro
            </span>
          </div>
          <div
            className="property__description"
          >
            <p
              className="property__text"
            >
              This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.
            </p>
          </div>
        </div>
        <section
          className="property__reviews reviews"
        >
          <h2
            className="reviews__title"
          >
            Reviews ·
            <span
              className="reviews__amount"
            >
              3
            </span>
          </h2>
          <ul
            className="reviews__list"
          >
            <li
              className="reviews__item"
            >
              <div
                className="reviews__user user"
              >
                <div
                  className="reviews__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    alt="Review avatar"
                    className="reviews__avatar user__avatar"
                    height="54"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/6.jpg"
                    width="54"
                  />
                </div>
                <span
                  className="reviews__user-name"
                >
                  Kendall
                </span>
              </div>
              <div
                className="reviews__info"
              >
                <div
                  className="reviews__rating rating"
                >
                  <div
                    className="reviews__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "80%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <p
                  className="reviews__text"
                >
                  We loved it so much, the house, the veiw, the location just great.. Highly reccomend :)
                </p>
                <time
                  className="reviews__time"
                  dateTime="2019-04-24"
                >
                  December 2019
                </time>
              </div>
            </li>
            <li
              className="reviews__item"
            >
              <div
                className="reviews__user user"
              >
                <div
                  className="reviews__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    alt="Review avatar"
                    className="reviews__avatar user__avatar"
                    height="54"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/5.jpg"
                    width="54"
                  />
                </div>
                <span
                  className="reviews__user-name"
                >
                  Corey
                </span>
                <span
                  className="reviews__user-status"
                >
                  Pro
                </span>
              </div>
              <div
                className="reviews__info"
              >
                <div
                  className="reviews__rating rating"
                >
                  <div
                    className="reviews__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "60%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <p
                  className="reviews__text"
                >
                  What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!
                </p>
                <time
                  className="reviews__time"
                  dateTime="2019-04-24"
                >
                  December 2019
                </time>
              </div>
            </li>
            <li
              className="reviews__item"
            >
              <div
                className="reviews__user user"
              >
                <div
                  className="reviews__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    alt="Review avatar"
                    className="reviews__avatar user__avatar"
                    height="54"
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/10.jpg"
                    width="54"
                  />
                </div>
                <span
                  className="reviews__user-name"
                >
                  Christina
                </span>
              </div>
              <div
                className="reviews__info"
              >
                <div
                  className="reviews__rating rating"
                >
                  <div
                    className="reviews__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "60%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <p
                  className="reviews__text"
                >
                  What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!
                </p>
                <time
                  className="reviews__time"
                  dateTime="2019-04-24"
                >
                  December 2019
                </time>
              </div>
            </li>
          </ul>
          <form
            action="#"
            className="reviews__form form"
            method="post"
          >
            <label
              className="reviews__label form__label"
              htmlFor="review"
            >
              Your review
            </label>
            <div
              className="reviews__rating-form form__rating"
            >
              <input
                className="form__rating-input visually-hidden"
                disabled={false}
                id="5-stars"
                name="rating"
                onChange={[Function]}
                type="radio"
                value={5}
              />
              <label
                className="reviews__rating-label form__rating-label"
                htmlFor="5-stars"
                title="perfect"
              >
                <svg
                  className="form__star-image"
                  height="33"
                  width="37"
                >
                  <use
                    xlinkHref="#icon-star"
                  />
                </svg>
              </label>
              <input
                className="form__rating-input visually-hidden"
                disabled={false}
                id="4-stars"
                name="rating"
                onChange={[Function]}
                type="radio"
                value={4}
              />
              <label
                className="reviews__rating-label form__rating-label"
                htmlFor="4-stars"
                title="good"
              >
                <svg
                  className="form__star-image"
                  height="33"
                  width="37"
                >
                  <use
                    xlinkHref="#icon-star"
                  />
                </svg>
              </label>
              <input
                className="form__rating-input visually-hidden"
                disabled={false}
                id="3-stars"
                name="rating"
                onChange={[Function]}
                type="radio"
                value={3}
              />
              <label
                className="reviews__rating-label form__rating-label"
                htmlFor="3-stars"
                title="not bad"
              >
                <svg
                  className="form__star-image"
                  height="33"
                  width="37"
                >
                  <use
                    xlinkHref="#icon-star"
                  />
                </svg>
              </label>
              <input
                className="form__rating-input visually-hidden"
                disabled={false}
                id="2-stars"
                name="rating"
                onChange={[Function]}
                type="radio"
                value={2}
              />
              <label
                className="reviews__rating-label form__rating-label"
                htmlFor="2-stars"
                title="badly"
              >
                <svg
                  className="form__star-image"
                  height="33"
                  width="37"
                >
                  <use
                    xlinkHref="#icon-star"
                  />
                </svg>
              </label>
              <input
                className="form__rating-input visually-hidden"
                disabled={false}
                id="1-stars"
                name="rating"
                onChange={[Function]}
                type="radio"
                value={1}
              />
              <label
                className="reviews__rating-label form__rating-label"
                htmlFor="1-stars"
                title="terribly"
              >
                <svg
                  className="form__star-image"
                  height="33"
                  width="37"
                >
                  <use
                    xlinkHref="#icon-star"
                  />
                </svg>
              </label>
            </div>
            <textarea
              className="reviews__textarea form__textarea"
              disabled={false}
              id="review"
              name="review"
              onChange={[Function]}
              placeholder="Tell how was your stay, what you like and what can be improved"
              value=""
            />
            <div
              className="reviews__button-wrapper"
            >
              <p
                className="reviews__help"
              >
                To submit review please make sure to set 
                <span
                  className="reviews__star"
                >
                  rating
                </span>
                 and describe your stay with at least 
                <b
                  className="reviews__text-amount"
                >
                  50 characters
                </b>
                .
              </p>
              <button
                className="reviews__submit form__submit button"
                disabled={true}
                onClick={[Function]}
                type="submit"
              >
                Submit
              </button>
            </div>
            
          </form>
        </section>
      </div>
    </div>
    <section
      className="property__map map"
    />
  </section>
  <div
    className="container"
  >
    <section
      className="near-places places"
    >
      <h2
        className="near-places__title"
      >
        Other places in the neighbourhood
      </h2>
      <div
        className="near-places__list places__list"
      />
    </section>
  </div>
</main>
`;
